### ISSUE TITLE
[BUG] Dynamic provisioning has reduced permissions on mounted volumes; unable to create directories or modify permissions

### ISSUE THREAD

**Issue by hippyod** [crc-org/crc#3440]
### General information

  * OS: Fedora 37
  * Hypervisor: KVM 
  * Did you run `crc setup` before starting it (Yes/No)? Yes
  * Running CRC on: Baremetal-Server

## CRC version
```bash
# Put `crc version` output here
CRC version: 2.11.0+823e40d
OpenShift version: 4.11.13
Podman version: 4.2.0
```
  
## CRC status
```bash
# Put `crc status --log-level debug` output here
DEBU CRC version: 2.11.0+823e40d                  
DEBU OpenShift version: 4.11.13                   
DEBU Podman version: 4.2.0                        
DEBU Running 'crc status'                         
CRC VM:          Running
OpenShift:       Running (v4.11.13)
RAM Usage:       12GB of 67.53GB
Disk Usage:      30.43GB of 136.9GB (Inside the CRC VM)
Cache Usage:     15.99GB
Cache Directory: /home/hippyod/.crc/cache
```

## CRC config
```bash
# Put `crc config view` output here
- consent-telemetry                     : no
```

## Host Operating System
```bash
# Put the output of `cat /etc/os-release` in case of Linux
NAME="Fedora Linux"
VERSION="37 (Workstation Edition)"
ID=fedora
VERSION_ID=37
VERSION_CODENAME=""
PLATFORM_ID="platform:f37"
PRETTY_NAME="Fedora Linux 37 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:37"
DEFAULT_HOSTNAME="fedora"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f37/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=37
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=37
VARIANT="Workstation Edition"
VARIANT_ID=workstation
```

### Steps to reproduce

  1. Create pvc
  2. mount to Jenkins build agent user home directory (or anything where write permissions are expected, really)
  3. try and create a directory in mounted volume, where said volume was given 0777 permission
  4. Watch if fail hard: 
```bash
helm repo add elCicdCharts https://raw.githubusercontent.com/elcicd/el-CICD-deploy/main/charts
Error: mkdir /home/jenkins/.config: permission denied
```

### Expected
User in mounted volume has write permissions.

### Actual
User doesn't have write permissions, or permission to change the permissions on a directory they should own.

### Logs
```bash
helm repo add elCicdCharts https://raw.githubusercontent.com/elcicd/el-CICD-deploy/main/charts
Error: mkdir /home/jenkins/.config: permission denied
```

Bottom line is the new dynamic volume strategy has changed the nature of  what's being created somehow.  I've been using crc's dynamic volumes for years in the manner above without issue, and it's broken and I have no idea why.  The just previous release worked fine, so this has nothing to do with 4.11 SCC changes (and I'm using a custom SCC in this case, regardless).  Only difference is the change to dynamic provisioning.  Thx.

**Comment by praveenkumar** [comment]
We recently changed the way of providing PVC on demand using dynamic provisioning https://github.com/crc-org/snc/commit/77aa37ee5a74b3cff7506bae3b53292d028073f9 , before we used to have preconfigured pv with `770` permission and now it handled by dynamic provisioner. We need to check what is missing in that config which causing this issue.

**Comment by hippyod** [comment]
@praveenkumar OK, I solved my issue at least.  For the Deployment spec I added the following in my Jenkins build agent:

```
apiVersion: "v1"
kind: "Pod"
...
spec:
  securityContext:
    fsGroup: 1001    
```

This fixed my particular problem, and everything behaves normally again (at least for my build agents, which run in a modified SCC).  This isn't correct though, as far as I am concerned.  Shouldn't I just have automatic write permissions?  I haven't tried this fix with a typical pod created under the default SCC yet, but when I have a chance I'll check out whether this issue can also be fixed similarly for those pods.  Hopefully this info and further tests will help you out what's going on.

I also noticed while searching for a solution that in minikube there are tricks you have to do to the host volume to get it writable when mounted in pods?  So maybe there's an issue between the preconfigured volumes you used and what the dynamic provisioner is allowing...?  You hinted at that, too, of course, but it's outside my developer oriented wheelhouse to figure it out.

Again, I wish to extend my heartfelt thanks with this fix.  Been waiting a long time for it.  With every fix I understand problems will arise until smoothed out.  Thanks for your help, and sorry for the bother.

**Comment by hippyod** [comment]
Oh, and there's also the issue I mentioned in the discussions, whereby PVC's now remain in Pending status and do not convert to Bound until attached to a Pod.  I found some workarounds, but that's probably not the behavior we want?

**Comment by praveenkumar** [comment]
@hippyod Thank you for try out and providing those useful info. We are also not expert around storage side and just followed the document to add this dynamic provision support, I think it would be much better if we have some integration tests (like pod/deployment resource which consume PV and get required behavior, and once delete the resource the pv should again goes to expected state).

**Comment by hippyod** [comment]
@praveenkumar OK, so after some more work and thanks to a very knowledgeable work colleague, default behavior for mounted volumes that are dynamically provisioned (using a BusyBox instance) is that the volume should be writable by the default container uid; i.e. owner and group on the mounted volume should be the uid OCP provides.  **The current solution is a breaking change.**

While there is the workaround I posted above, I don't want to have to pollute the code in my Helm charts and CICD pipelines just to keep things working, so I cannot upgrade until this is fixed.  This is actually much, much worse than the previous behavior where recycling wouldn't work (at least for me), which was an extreme annoyance but at least everything behaves as intended from the perspective of my software while there are still PV's available. 

My investigation (so far, and I am not very good in this area) thinks that there isn't an easy fix using the CSI driver you've chosen.  If I have time over the next few days I will try and help and investigate some potential replacements solutions, including Red Hat's Local Volume operator.  If I have any success I will post the solutions here to discuss.

Like I said, I can't upgrade until this is fixed.  I can't have my code base polluted with an unnecessary workaround.  From my perspective, if a reasonable and maintainable solution can't be found in a reasonable amount of time, I'd vote for going back the pre-provisioned PV's and fixing that solution until a better and correct one can be found.

Thanks for your attention to this.  Really appreciate it.  Sorry for the bother.

**Comment by hippyod** [comment]
@praveenkumar [NOTE: deleted a previous comment as unhelpful.]

OK, been working on this day and night (actually, kind of literally), and I think I figured out what going on.  When I deployed Jenkins, I was getting the right group id on all mounted volumes:
```
sh-5.1$ ls -al
total 104
drwxrwsr-x.  20 root       1000850000  4096 Dec  4 21:46 .
drwxr-xr-x.   1 root       root         222 Dec  4 09:53 ..
drwxr-sr-x.   5 1000850000 1000850000    72 Dec  4 21:31 .cache
drwxr-sr-x.   3 1000850000 1000850000    18 Dec  4 21:33 .config
...
```
But on the Jenkins Agent, I was getting root as the default gid:
```
sh-5.2$ ls -al
total 0
drwxr-xr-x. 3 root    root  19 Dec  4 21:34 .
drwxrwsrwx. 1 jenkins root  21 Dec  4 20:33 ..
```
 which was breaking the Jenkins agents.  I couldn't figure out what was going on until I remembered that I auto-magically set a default `securityContext` on the `PodTemplate`  [NOTE: **not** the container `securityContext`] I create via Helm chart due to the potentially breaking updated security policy changes introduced in OCP4.11:
```
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault

```
What happens when the Pod  is deployed, the actual `securityContext` is realized as:
```
  securityContext:
    fsGroup: 1000850000
    runAsNonRoot: true
    seLinuxOptions:
      level: s0:c29,c19
    seccompProfile:
      type: RuntimeDefault
```
Notice how `fsGroup` is properly set now?  Which is why there's no problem.  On the Jenkins Agent, on the other hand, the Jenkins Kubernetes Plugin is managing the deployment, rather than my Helm chart,  and since it's only worried about vanilla k8s, it just deploys with an empty `securityContext`.  OCP doesn't inject the `fsGroup` in this case, so the volume is mounted in RO mode (root gid).

Bottom line is the following:

a) **Dynamic provisioning works as it should.  There is no bug or breaking change.**

b) `spec.template.spec.securityContext` **should always be explicitly defined if you want your mounted volumes to behave properly.** As a lot of us have found out, regardless of volumes or not, defining your `securityContext` in 4.11 for a `PodTemplate` needs to become a best practice.

Totally apologize for bothering you.  Hope this helps.
